stratosphere/ncm/source/ncm_content_meta_database.cpp

```diff
@@ -0,0 +1,584 @@
+                        default:
+                            /* The first u64 of all non-application extended headers is the application program id. */
+                            entry_application_id = *GetValueExtendedHeader<ProgramId>(value);
+                    }
+
+                    /* Application id doesn't match filter, skip this entry. */
+                    if (entry_application_id != application_program_id) {
+                        continue;
+                    }
+                }
+            }
+
+            /* Write the entry to the output buffer. */
+            if (entries_written < out_info.GetSize()) {
+                out_info[entries_written] = key;
+                entries_written++;
+            }
+
+            entries_total++;
+        }
+
+        out_entries_total.SetValue(entries_total);
+        out_entries_written.SetValue(entries_written);
+        return ResultSuccess();
+    }
+
+    Result ContentMetaDatabaseInterface::GetLatestContentMetaKey(sf::Out<ContentMetaKey> out_key, ProgramId program_id) {   
+        R_TRY(this->EnsureEnabled());
+        ContentMetaKey key;        
+        R_TRY(this->GetLatestContentMetaKeyImpl(&key, program_id));
+        out_key.SetValue(key);
+        return ResultSuccess();
+    }
+
+    Result ContentMetaDatabaseInterface::ListApplication(sf::Out<u32> out_entries_total, sf::Out<u32> out_entries_written, const sf::OutArray<ApplicationContentMetaKey> &out_keys, ContentMetaType type) {
+        R_TRY(this->EnsureEnabled());
+        
+        size_t entries_total = 0;
+        size_t entries_written = 0;
+
+        /* If there are no entries then we've already successfully listed them all. */
+        if (this->kvs->GetCount() == 0) {
+            out_entries_total.SetValue(entries_total);
+            out_entries_written.SetValue(entries_written);
+            return ResultSuccess();
+        }
+
+        for (auto entry = this->kvs->begin(); entry != this->kvs->end(); entry++) {
+            ContentMetaKey key = entry->GetKey();
+
+            /* Check if this entry matches the given filters. */
+            if (!((static_cast<u8>(type) == 0 || key.type == type))) {
+                continue;
+            }
+
+            const void* value = nullptr;
+            size_t value_size = 0;
+            R_TRY(GetContentMetaValuePointer(&value, &value_size, key, this->kvs));
+
+            if (key.type == ContentMetaType::Application || key.type == ContentMetaType::Patch || key.type == ContentMetaType::AddOnContent || key.type == ContentMetaType::Delta) {
+                ProgramId application_id = key.id;
+                
+                switch (key.type) {
+                    case ContentMetaType::Application:
+                        break;
+                    default:
+                        /* The first u64 of all non-application extended headers is the application program id. */
+                        application_id = *GetValueExtendedHeader<ProgramId>(value);
```

Comment:
replace switch as discussed.

Also the logic here is deeply similar to above, needs reader pattern but I'll probably end up handling that.